odc/geo/_map.py

```diff
@@ -153,3 +153,141 @@ def add_to(
+        to plot on the map.
+    :param map:
+        An optional existing :py:mod:`folium` map object to plot into.
+        By default, a new map object will be created.
+    :param bands:
+        Bands used for RGB colours when converting from a
+        :py:class:`~xarray.Dataset` (order should be red, green, blue).
+        By default, the function will attempt to guess bands
+        automatically. Ignored for :py:class:`~xarray.DataArray` inputs.
+    :param vmin:
+        Lower value used for the color stretch.
+    :param vmax:
+        Upper value used for the color stretch.
+    :param cmap:
+        The colormap used to colorise single-band arrays. If not
+        provided, this will default to 'viridis'. Ignored for multi-band
+        inputs.
+    :param robust:
+        If ``True`` (and ``vmin`` and ``vmax`` are absent), the colormap
+        range will be computed based on 2nd and 98th percentiles,
+        minimising the influence of extreme values. Used for single-band
+        arrays only; ignored for multi-band inputs.
+    :param tiles:
+        Map tileset to use for the map basemap. Supports any option
+        supported by :py:mod:`folium`, including "OpenStreetMap",
+        "CartoDB positron", "CartoDB dark_matter" or a custom XYZ URL.
+    :param attr:
+        Map tile attribution; only required if passing custom tile URL.
+    :param layer_control:
+        Whether to add a control to the map to show or hide map layers.
+        If a layer control already exists, this will be skipped.
+    :param resampling:
+        Custom resampling method to use when reprojecting ``xx`` to the
+        map CRS; defaults to "nearest".
+    :param map_kwds:
+        Additional keyword arguments to pass to ``folium.Map()``.
+    :param reproject_kwds:
+        Additional keyword arguments to pass to ``.odc.reproject()``.
+    :param \**kwargs:
+        Additional keyword arguments to pass to ``.odc.add_to()``.
+
+    :return: A :py:mod:`folium` map containing the plotted xarray data.
+    """
+    assert have.folium
+
+    from folium import Map, LayerControl
+
+    # Empty kwargs by default
+    if map_kwds is None:
+        map_kwds = {}
+    if reproject_kwds is None:
+        reproject_kwds = {}
+
+    # Update any supplied kwargs with custom params
+    kwargs.update(cmap=cmap, vmin=vmin, vmax=vmax, robust=robust)
+    map_kwds.update(tiles=tiles, attr=attr)
+    reproject_kwds.update(resampling=resampling)
+
+    # If input is a dataset, convert to an RGBA array
+    if isinstance(xx, xr.Dataset):
+        try:
+            xx = xx.odc.to_rgba(bands=bands, vmin=vmin, vmax=vmax)
+        except ValueError as e:
+            raise ValueError(
+                f"Unable to automatically guess RGB colours ({e}). "
+                f"Manually specify bands to plot using the `bands` "
+                "parameter, e.g. `ds.odc.explore(bands=['a', 'b', 'c'])`"
+            ) from e
```

Comment:
`.add_to` does all of that and does it after all the projection and max size shrinkage. Code here should not perform any colorization and projection change, instead it should call `.add_to` with appropriate parameters.

Reply:
Thanks @Kirill888! Looks like I can simplify this a fair bit - I didn't realise that `add_to` did its own reprojection. I did however reimplement some of the error catching (e.g. too many dims) outside of `.add_to` was that I was finding it would only raise errors _after_ reprojection and loading into memory - which would cause issues if users passed in e.g. hundreds of timesteps instead of one (data would load slowly then only return the error at the end). I'll have another look and see if I was doing something wrong to cause that to happen.

A few quick questions:
1. I would like to be able to support custom resampling when adding to the map, but [`.add_to` doesn't seem to support this](https://github.com/opendatacube/odc-geo/blob/develop/odc/geo/_map.py#L138). Any objection to me adding a `resampling` param to `.add_to`?
2. Does `.add_to` automatically convert `xr.Dataset` inputs to an RGBA array? I can't see `.add_to` as a method on `xr.Dataset`, only [xr.DataArrays](https://github.com/opendatacube/odc-geo/blob/a9fbfc53625502d14dc7f124470ae0d54eb52f60/odc/geo/_xr_interop.py#L851) (I'd like `.explore()` to automatically convert and plot `xr.Dataset` as multiband RGB, and `xr.DataArray` inputs to plot as single-band colorised arrays).
3. What is the advantage of running `.add_to(..., map=None)`, rather than the current approach of simply creating the map first and doing `.add_to(..., map=m)`?

Reply:
Extra features like resampling config, or more robust input verification should go into `.add_to`.

Reply:
I've worked out the problem with Dask inputs: if you accidently pass in an array with e.g. hundreds of timesteps, it gets loaded into memory inside `compress` here:
https://github.com/opendatacube/odc-geo/blob/develop/odc/geo/_compress.py#L95-L96

But that happens before the excess time dimension is caught in `_compress_image` here:
https://github.com/opendatacube/odc-geo/blob/develop/odc/geo/_compress.py#L26-L33

That causes a long wait for data to load, with users only receiving the "Expect 2 or 3 dimensional array" error at the very end. Ideally, we'd be able to raise that error much earlier, avoiding the need to wait until the end of a potentially large Dask load.

Reply:
> I've worked out the problem with Dask inputs: if you accidently pass in an array with e.g. hundreds of timesteps, it gets loaded into memory inside `compress` here: https://github.com/opendatacube/odc-geo/blob/develop/odc/geo/_compress.py#L95-L96

So ideally compress should catch that before calling "compute", something like `_verify_can_compress(xx: xr.DataArray)` that should contain logic similar to:

https://github.com/opendatacube/odc-geo/blob/a9fbfc53625502d14dc7f124470ae0d54eb52f60/odc/geo/_compress.py#L23-L33

which is basically: array `.ndim in (2,3)` after squeezing out single sized dimensions.

Reply:
> 2\. Does `.add_to` automatically convert `xr.Dataset` inputs to an RGBA array? I can't see `.add_to` as a method on `xr.Dataset`, only [`xr.DataArray`](https://github.com/opendatacube/odc-geo/blob/a9fbfc53625502d14dc7f124470ae0d54eb52f60/odc/geo/_xr_interop.py#L851) (I'd like `.explore()` to work on both: automatically converting and plotting `xr.Dataset` as an multiband RGB image, and  plotting `xr.DataArray`s as a single-band colorised image).

I see, let's keep `.add_to` as is then, and perform `Dataset+bands->DataArray` conversion inside `.explore`.

Reply:
Sounds good, will put something together 👍